File: IDEA/src/ksp/kos/ideaplugin/reference/context/FileContext.java
package ksp.kos.ideaplugin.reference.context;

import ksp.kos.ideaplugin.KerboScriptFile;
import ksp.kos.ideaplugin.dataflow.ReferenceFlow;
import ksp.kos.ideaplugin.reference.ReferableType;
import ksp.kos.ideaplugin.reference.Reference;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Created on 08/10/16.
 *
 * @author ptasha
 */
public abstract class FileContext extends LocalContext implements ReferenceFlow, FileDuality {
    private final String name;

    public FileContext(LocalContext parent, String name, FileContextResolver fileResolver) {
        this(parent, name, createResolvers(fileResolver));
    }

    protected FileContext(LocalContext parent, String name, List<ReferenceResolver<LocalContext>> resolvers) {
        super(parent, resolvers);
        this.name = name;
    }

    public static List<ReferenceResolver<LocalContext>> createResolvers(FileContextResolver fileResolver) {
        ArrayList<ReferenceResolver<LocalContext>> resolvers = new ArrayList<>();
        resolvers.add(new FileResolver(fileResolver));
        resolvers.add(new LocalResolver());
        resolvers.add(new VirtualResolver());
        resolvers.add(new ImportsResolver(fileResolver));
        return resolvers;
    }

    @Override
    public LocalContext getKingdom() {
        return this;
    }

    @Override
    public ReferableType getReferableType() {
        return ReferableType.FILE;
    }

    @Override
    public String getName() {
        return name;
    }

    public Map<String, Duality> getImports() {
        return getDeclarations(ReferableType.FILE);
    }

    @Override
    protected void registerUnknown(ReferableType type, String name, Duality element) {
        if (type==ReferableType.FILE) {
            addDefinition(type, KerboScriptFile.stripExtension(name), element);
        } else {
            super.registerUnknown(type, name, element);
        }
    }

    public static class FileResolver implements ReferenceResolver<LocalContext> {

        private final FileContextResolver fileContextResolver;

        public FileResolver(FileContextResolver fileContextResolver) {
            this.fileContextResolver = fileContextResolver;
        }

        @Override
        public Duality resolve(LocalContext context, Reference reference, boolean createAllowed) {
            if (reference.getReferableType()==ReferableType.FILE) {
                return fileContextResolver.resolveFile(reference.getName());
            }
            return null;
        }
    }

    @Override
    public FileContext getSemantics() {
        return this;
    }

    public static class ImportsResolver implements ReferenceResolver<LocalContext> {

        private final FileContextResolver fileContextResolver;

        public ImportsResolver(FileContextResolver fileContextResolver) {
            this.fileContextResolver = fileContextResolver;
        }

        @Override
        public Duality resolve(LocalContext context, Reference reference, boolean createAllowed) {
            for (Duality run : context.getDeclarations(ReferableType.FILE).values()) {
                FileDuality dependency = fileContextResolver.resolveFile(run.getName());
                if (dependency != null) {
                    Duality resolved = dependency.getSemantics().findLocalDeclaration(reference);
                    if (resolved != null) {
                        return resolved;
                    }
                }
            }
            return null;
        }
    }

    private static class VirtualResolver extends ParentResolver {
        @Override
        public Duality resolve(LocalContext context, Reference reference, boolean createAllowed) {
            if (!createAllowed) {
                return null;
            }
            return super.resolve(context, reference, createAllowed);
        }
    }
}

File: IDEA/src/main/java/ksp/kos/ideaplugin/reference/context/FileContext.kt
package ksp.kos.ideaplugin.reference.context

import com.intellij.openapi.project.ProjectManager
import com.intellij.openapi.vfs.VfsUtilCore
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.VirtualFileManager
import com.intellij.psi.PsiManager
import ksp.kos.ideaplugin.KerboScriptFile
import ksp.kos.ideaplugin.dataflow.ReferenceFlow
import ksp.kos.ideaplugin.reference.OccurrenceType
import ksp.kos.ideaplugin.reference.ReferableType
import ksp.kos.ideaplugin.reference.Reference
import java.net.URL

/**
 * Created on 08/10/16.
 *
 * @author ptasha
 */
abstract class FileContext protected constructor(
    parent: LocalContext?,
    private val name: String,
    resolvers: List<ReferenceResolver<LocalContext>>,
) : LocalContext(parent, resolvers), ReferenceFlow<FileContext>, FileDuality {

    constructor(parent: LocalContext?, name: String, fileResolver: FileContextResolver)
            : this(parent, name, createResolvers(fileResolver))

    override fun getKingdom(): LocalContext = this

    override fun getReferableType(): ReferableType = ReferableType.FILE

    override fun getName(): String = name

    val imports: Map<String, Duality>
        get() = getDeclarations(ReferableType.FILE)

    override fun registerUnknown(type: ReferableType, name: String, element: Duality) {
        if (type == ReferableType.FILE) {
            addDefinition(type, KerboScriptFile.stripExtension(name), element)
        } else {
            super.registerUnknown(type, name, element)
        }
    }

    class FileResolver(private val fileContextResolver: FileContextResolver) : ReferenceResolver<LocalContext> {
        override fun resolve(context: LocalContext, reference: Reference, createAllowed: Boolean): Duality? =
            if (reference.referableType == ReferableType.FILE) {
                fileContextResolver.resolveFile(reference.name)
            } else {
                null
            }
    }

    override fun getSemantics(): FileContext = this

    class ImportsResolver(private val fileContextResolver: FileContextResolver) : ReferenceResolver<LocalContext> {
        override fun resolve(context: LocalContext, reference: Reference, createAllowed: Boolean): Duality? =
            context.getDeclarations(ReferableType.FILE)
                .values
                .mapNotNull { run ->
                    fileContextResolver.resolveFile(run.name)
                        ?.semantics
                        ?.findLocalDeclaration(reference, OccurrenceType.GLOBAL)
                }
                .firstOrNull()
    }

    class BuiltinResolver : ReferenceResolver<LocalContext> {
        override fun resolve(context: LocalContext, reference: Reference, createAllowed: Boolean): Duality? {
            if (ksFile == null) {
                tryGetBuiltinKsFile()
            }
            return ksFile?.semantics?.findLocalDeclaration(reference, OccurrenceType.GLOBAL)
        }

        companion object {
            private var ksFile: KerboScriptFile? = null

            private fun URL.toVirtualFile(): VirtualFile? {
                val urlStr = VfsUtilCore.convertFromUrl(this)
                return VirtualFileManager.getInstance().findFileByUrl(urlStr)
            }

            private fun tryGetBuiltinKsFile() {
                val builtinVirtualFile = BuiltinResolver::class.java.classLoader
                    .getResource("builtin.ks")
                    ?.toVirtualFile()
                    ?: return

                val project = ProjectManager.getInstance().defaultProject
                val psiFile = PsiManager.getInstance(project).findFile(builtinVirtualFile)
                ksFile = psiFile as? KerboScriptFile
            }
        }
    }

    private class VirtualResolver : ParentResolver() {
        override fun resolve(context: LocalContext, reference: Reference, createAllowed: Boolean): Duality? =
            if (!createAllowed) null else super.resolve(context, reference, createAllowed)
    }

    companion object {
        @JvmStatic
        fun createResolvers(fileResolver: FileContextResolver): MutableList<ReferenceResolver<LocalContext>> =
            mutableListOf(
                FileResolver(fileResolver),
                LocalResolver(),
                VirtualResolver(),
                ImportsResolver(fileResolver),
                BuiltinResolver(), // Make sure this is last, to allow shadowing of builtin values
            )
    }
}
